pageserver/src/tenant_mgr.rs

```diff
@@ -104,6 +105,64 @@ struct Tenant {
     local_timelines: HashMap<ZTimelineId, Arc<DatadirTimelineImpl>>,
 }
 
+impl Tenant {
+    fn get_local_timeline_with_load(
+        &mut self,
+        timeline_id: ZTimelineId,
+    ) -> anyhow::Result<Arc<DatadirTimelineImpl>> {
+        if let Some(page_tline) = self.local_timelines.get(&timeline_id) {
+            Ok(Arc::clone(page_tline))
+        } else {
+            let page_tline = self
+                .load_local_timeline(timeline_id)
+                .with_context(|| format!("Failed to load local timeline {timeline_id}"))?;
+            self.local_timelines
+                .insert(timeline_id, Arc::clone(&page_tline));
+            Ok(page_tline)
+        }
+    }
+
+    fn load_local_timeline(
+        &mut self,
+        timeline_id: ZTimelineId,
+    ) -> anyhow::Result<Arc<DatadirTimeline<LayeredRepository>>> {
+        let inmem_timeline = self.repo.get_timeline_load(timeline_id).with_context(|| {
+            format!("Inmem timeline {timeline_id} not found in tenant's repository")
+        })?;
+        let repartition_distance = self.repo.get_checkpoint_distance() / 10;
+        let mut ancestor_logical_size: Option<usize> = None;
+        if let Some(ancestor_id) = inmem_timeline.get_ancestor_timeline_id() {
+            if let Ok(ancestor_pgdir) = self.get_local_timeline_with_load(ancestor_id) {
+                if ancestor_pgdir.tline.get_last_record_lsn() == inmem_timeline.get_ancestor_lsn() {
+                    ancestor_logical_size = Some(ancestor_pgdir.get_current_logical_size());
+                    debug!("Branching: copy ancestor logical size");
+                } else {
+                    debug!(
+                        "Brancing: ancestor LSN={}, branch LSN={}",
+                        ancestor_pgdir.tline.get_last_record_lsn(),
+                        inmem_timeline.get_ancestor_lsn()
+                    );
+                }
+            }
+        };
+        let pgdir = Arc::new(DatadirTimelineImpl::new(
```

Comment:
nit: for consistency with the `load_local_timeline`, this should be named `page_tline` or `tline`. Not sure if `pgdir` fits the variable

Reply:
Ehhh... This is really mess with all this different timelines which are not actually timelines (doesn't implement `Timeline` trait. So `page_tline`seems to be very confusing (at least for me): it is not strictly a timeline and why "page"?